File: 1erdepartamental/python/21.quick_indirecto.py
# ============================================
# PROGRAMA: Quick Sort Indirecto
# DESCRIPCIÓN: Quick Sort ordenando índices
# ============================================

from abc import ABC, abstractmethod


class OrdenadorIndirectoBase(ABC):
    """Base para ordenadores indirectos"""

    def __init__(self, datos):
        self.__datos = datos
        self.__indices = list(range(len(datos)))

    @property
    def datos_originales(self):
        return self.__datos

    @property
    def indices(self):
        return self.__indices.copy()

    @abstractmethod
    def _ordenar_indices(self, idx):
        pass

    def ordenar(self):
        self.__indices = self._ordenar_indices(self.__indices)
        return self.__indices

    def acceder_ordenado(self):
        return [self.__datos[i] for i in self.__indices]


class OrdenadorQuickSortIndirecto(OrdenadorIndirectoBase):
    """Quick Sort que ordena índices"""

    def _ordenar_indices(self, idx):
        if len(idx) <= 1:
            return idx

        # Pivote: índice del medio
        pivote_val = self._OrdenadorIndirectoBase__datos[idx[len(idx) // 2]]

        # Particiona índices según su valor en datos
        menores = [
            i for i in idx if self._OrdenadorIndirectoBase__datos[i] < pivote_val
        ]
        iguales = [
            i for i in idx if self._OrdenadorIndirectoBase__datos[i] == pivote_val
        ]
        mayores = [
            i for i in idx if self._OrdenadorIndirectoBase__datos[i] > pivote_val
        ]

        return self._ordenar_indices(menores) + iguales + self._ordenar_indices(mayores)


class Persona:
    def __init__(self, nombre, edad):
        self.__nombre = nombre
        self.__edad = edad

    @property
    def nombre(self):
        return self.__nombre

    @property
    def edad(self):
        return self.__edad

    def __lt__(self, otro):
        return self.__edad < otro.edad

    def __eq__(self, otro):
        if not isinstance(otro, Persona):
            return False
        return self.__edad == otro.edad

    def __str__(self):
        return f"{self.__nombre}({self.__edad})"


class OrdenadorQuickSortIndirectoPersonas(OrdenadorQuickSortIndirecto):
    pass


# ============================================
# EJECUCIÓN
# ============================================
if __name__ == "__main__":
    datos = [64, 34, 25, 12, 22, 11, 90]
    o = OrdenadorQuickSortIndirecto(datos)
    o.ordenar()

    print(f"Datos originales: {datos}")
    print(f"Acceso ordenado: {o.acceder_ordenado()}")

    print()

    personas = [Persona("Juan", 25), Persona("Ana", 30), Persona("Pedro", 20)]
    op = OrdenadorQuickSortIndirectoPersonas(personas)
    op.ordenar()
    print(f"Por edad: {[str(p) for p in op.acceder_ordenado()]}")

File: 1erdepartamental/c++/21.quick_indirecto.cpp
// ============================================
// PROGRAMA: Quick Sort Indirecto
// DESCRIPCIÓN: Ordena índices con particionamiento
// ============================================

#include <iostream>
#include <vector>
using namespace std;


/**
 * Quick sort indirecto: ordena índices, preserva original
 */
template<typename T>
class OrdenadorIndirectoQuick {
private:
    const vector<T>* __datosOriginales;
    vector<int> __indices;
    
    int particionar(int bajo, int alto) {
        // Usa el valor original del último índice como pivote
        T pivote = (*__datosOriginales)[__indices[alto]];
        int i = bajo - 1;
        
        for (int j = bajo; j < alto; j++) {
            // Compara valores originales usando índices
            if ((*__datosOriginales)[__indices[j]] < pivote) {
                i++;
                swap(__indices[i], __indices[j]);
            }
        }
        swap(__indices[i + 1], __indices[alto]);
        return i + 1;
    }
    
    void quickSortRecursivo(int bajo, int alto) {
        if (bajo >= alto) return;
        
        int pi = particionar(bajo, alto);
        
        quickSortRecursivo(bajo, pi - 1);
        quickSortRecursivo(pi + 1, alto);
    }

public:
    OrdenadorIndirectoQuick(const vector<T>& datos)
        : __datosOriginales(&datos) {
        for (size_t i = 0; i < datos.size(); i++) {
            __indices.push_back(i);
        }
    }
    
    void ordenar() {
        if (__indices.size() > 1) {
            quickSortRecursivo(0, __indices.size() - 1);
        }
    }
    
    vector<int> getIndices() const { return __indices; }
    
    void mostrarOrdenado() const {
        cout << "Orden mediante índices: ";
        for (int idx : __indices) {
            cout << (*__datosOriginales)[idx] << " ";
        }
        cout << endl;
    }
    
    void mostrarOriginal() const {
        cout << "Array original: ";
        for (const T& val : *__datosOriginales) {
            cout << val << " ";
        }
        cout << endl;
    }
};


// ============================================
// EJECUCIÓN
// ============================================
int main() {
    vector<int> numeros = {64, 34, 25, 12, 22, 11, 90};
    
    OrdenadorIndirectoQuick<int> ordenador(numeros);
    
    ordenador.mostrarOriginal();
    ordenador.ordenar();
    ordenador.mostrarOrdenado();
    ordenador.mostrarOriginal();
    
    return 0;
}
